core/storage/config/gae_models_test.py

```diff
@@ -195,10 +208,14 @@ def test_commit_with_updated_rules(self) -> None:
         )
         param_model.commit(feconf.SYSTEM_COMMITTER_ID, 'commit message', [])
 
-        new_rules = [
+        new_rules: List[platform_parameter_domain.PlatformParameterRuleDict] = [
             {
                 'filters': [
-                    {'type': 'app_version', 'value': '>1.2.3'}
+                    # Here, we are providing 'value' key but there is no key
```

Comment:
I found that `value` is present as a key for `filters`.
can you take a look at the docstring of this [function](https://github.com/oppia/oppia/blob/1985420396ee8a0007c1d9407e9627e52113e067/core/storage/config/gae_models.py#L145).

Reply:
Please check with @BenHenning what is the expected behaviour here. He is responsible for the platform parameter implementation.

I suspect the docstring and some tests are is incorrect, but probably worth confirming.

(Also, it's great that you found this error using types! But let's fix it rather than "paper over" it.)

Reply:
I believe this is code is wrong. The correct implementation should be:

```python
                    {'type': 'app_version', 'conditions': [('>', '1.2.3')]} 
```

I think there are examples of tests which set this up correctly, so I suspect that @seanlip is correct and there are a few places that are out-of-date (including, it seems, some of the documentation).

Good find.

Reply:
Thanks, @BenHenning, and @seanlip for clearing the ambiguity.  Docstring and test cases are changed.

Reply:
I just noticed that in many test cases there is a mismatch in fetched and input dictionaries. By this I mean:

If we have a dictionary like the below:
```
        input_new_rules: List[platform_parameter_domain.PlatformParameterRuleDict] = [
            {
                'filters': [
                    {'type': 'app_version', 'conditions': [('>', '1.2.3')]}
                ],
                'value_when_matched': True
            },
            {'filters': [], 'value_when_matched': False},
        ]
```

And if commit this by using the `.commit()` method of model OR if we convert it to JSON. Then the value of `conditions` 's key is changed to `'conditions': [['>', '1.2.3']]`. So, this converted dictionary looks like the below:
```
        output_new_rules: List[platform_parameter_domain.PlatformParameterRuleDict] = [
            {
                'filters': [
                    {'type': 'app_version', 'conditions': [['>', '1.2.3']]}
                ],
                'value_when_matched': True
            },
            {'filters': [], 'value_when_matched': False},
        ]
```
So, due to difference in output dict and input dict backend test fails. 


Now, my concern is that if we want to keep passing the test, we have to change the input dict's `condition`'s key value to `[['>', '1.2.3']]`. But if we change the value of `conditions` from tuple to list then mypy complaints about inappropriate type.
because expected type of `conditions` is List[Tuple[str, str]]

For reference: https://github.com/sahiljoster32/oppia/blob/128d0089ccbeac5b5628dc4aa068af71e919be79/core/domain/platform_parameter_domain_test.py#L2148

Also you can take a look at failing backend tests!!

Reply:
@seanlip PTAL at this case also.

Reply:
Lets do 2

Reply:
Done!